grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -256,4 +253,67 @@ class PrepaymentInvoiceService {
         // to generate invoice item partially for the same invoice
         return relatedObject?.invoiceItems?.find { InvoiceItem it -> it.inverse && it.invoice == invoice }
     }
+
+    /**
+     * Bulk editing invoice items quantity on final invoice for prepaid POs.
+     * We are allowing to edit only the quantity here. After editing invoice items quantity we have to
+     * change quantity on the corresponding inverse item.
+     *
+     * */
+    void updateItemsQuantity(String invoiceId, List items) {
+        Invoice invoice = Invoice.get(invoiceId)
+        if (!invoice) {
+            throw new IllegalArgumentException("Cannot find invoice with id: ${invoiceId}")
+        }
+        if (invoice.isPrepaymentInvoice || !invoice.hasPrepaymentInvoice) {
+            throw new IllegalArgumentException("Cannot update quantities on prepayment or non prepaid invoices")
+        }
+
+        items.each { Map item ->
+            updateInvoiceItemQuantity(item.id, item.quantity)
+        }
+    }
+
+    /**
+     * Updates invoice item's quantity and amount and finds related inverse item and updates quantity and amount
+     * */
+    void updateInvoiceItemQuantity(String itemId, Integer quantity) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (!invoiceItem) {
+            throw new IllegalArgumentException("Cannot find invoice with id: ${itemId}")
+        }
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("Quantity to change needs to be higher than 0")
+        }
+        if (!invoiceItem?.shipmentItem) {
+            throw new IllegalArgumentException("Invoice item is missing shipment item. Cannot edit quantity.")
+        }
+        Integer quantityAvailableToInvoice = invoiceItem.shipmentItem.quantityToInvoice
+        if (quantity > quantityAvailableToInvoice + invoiceItem.quantity) {
+            throw new IllegalArgumentException("Cannot update quantity to higher value than available to invoice")
+        }
```

Comment:
I asked a similar question in the other PR, but what about cancelled items? You shouldn't be able to update the invoice quantity right? Is that already handled by the quantityToInvoice field? Would it just be 0 in that case?